File: sdk/go/darkauth.go
// Package darkauth provides the official DARK-AUTH Go SDK v2.0.0.
package darkauth

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"runtime"
	"strings"
	"time"
)

const Version = "2.0.0"

// DarkAuth is the main client for the DARK-AUTH API.
type DarkAuth struct {
	AppID        string
	Secret       string
	Version      string
	ApiURL       string
	SessionToken string
	HWID         string
	HTTPClient   *http.Client
	User         map[string]interface{}
	LicenseInfo  map[string]interface{}
}

// Response represents a DARK-AUTH API response.
type Response struct {
	Success      bool                   `json:"success"`
	Message      string                 `json:"message"`
	Code         string                 `json:"code"`
	SessionToken string                 `json:"session_token,omitempty"`
	Level        int                    `json:"level,omitempty"`
	Status       string                 `json:"status,omitempty"`
	Value        string                 `json:"value,omitempty"`
	ExpiresAt    string                 `json:"expires_at,omitempty"`
	Update       map[string]interface{} `json:"update,omitempty"`
	Messages     []ChatMessage          `json:"messages,omitempty"`
	Extra        map[string]interface{} `json:"-"`
}

// ChatMessage represents a single chat message.
type ChatMessage struct {
	Sender    string `json:"sender"`
	Message   string `json:"message"`
	Channel   string `json:"channel,omitempty"`
	Timestamp string `json:"timestamp,omitempty"`
}

func (r *Response) UnmarshalJSON(data []byte) error {
	type Alias Response
	aux := &struct{ *Alias }{(*Alias)(r)}
	if err := json.Unmarshal(data, &aux); err != nil {
		return err
	}
	r.Extra = make(map[string]interface{})
	json.Unmarshal(data, &r.Extra)
	delete(r.Extra, "success")
	delete(r.Extra, "message")
	delete(r.Extra, "code")
	delete(r.Extra, "session_token")
	delete(r.Extra, "level")
	delete(r.Extra, "status")
	delete(r.Extra, "value")
	delete(r.Extra, "expires_at")
	delete(r.Extra, "update")
	delete(r.Extra, "messages")
	return nil
}

// New creates a new DARK-AUTH client.
func New(appID, secret, apiURL, version string) *DarkAuth {
	if version == "" {
		version = Version
	}

	hostname, _ := os.Hostname()
	raw := fmt.Sprintf("%s:%s:%s:%s:%d", hostname, os.Getenv("USER"), runtime.GOOS, runtime.GOARCH, os.Getpid())
	h := sha256.Sum256([]byte(raw))
	hwid := hex.EncodeToString(h[:])

	return &DarkAuth{
		AppID:    appID,
		Secret:   secret,
		Version:  version,
		ApiURL:   strings.TrimRight(apiURL, "/"),
		HWID:     hwid,
		HTTPClient: &http.Client{Timeout: 15 * time.Second},
	}
}

func (da *DarkAuth) url(endpoint string) string {
	return fmt.Sprintf("%s/api/v2/%s", da.ApiURL, strings.TrimLeft(endpoint, "/"))
}

func (da *DarkAuth) post(endpoint string, payload interface{}) (*Response, error) {
	b, err := json.Marshal(payload)
	if err != nil {
		return nil, err
	}

	resp, err := da.HTTPClient.Post(da.url(endpoint), "application/json", bytes.NewBuffer(b))
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var res Response
	if err := json.Unmarshal(body, &res); err != nil {
		return nil, err
	}

	if !res.Success {
		return nil, fmt.Errorf("[%s] %s", res.Code, res.Message)
	}

	return &res, nil
}

func (da *DarkAuth) get(endpoint string, params map[string]string) (*Response, error) {
	req, err := http.NewRequest("GET", da.url(endpoint), nil)
	if err != nil {
		return nil, err
	}

	q := req.URL.Query()
	for k, v := range params {
		q.Set(k, v)
	}
	req.URL.RawQuery = q.Encode()

	resp, err := da.HTTPClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var res Response
	if err := json.Unmarshal(body, &res); err != nil {
		return nil, err
	}

	return &res, nil
}

func (da *DarkAuth) ensureSession() error {
	if da.SessionToken == "" {
		_, err := da.Init("")
		return err
	}
	return nil
}

// Init initializes a session with the DARK-AUTH API.
func (da *DarkAuth) Init(binaryHash string) (*Response, error) {
	payload := map[string]interface{}{
		"app_id":  da.AppID,
		"secret":  da.Secret,
		"version": da.Version,
	}
	if binaryHash != "" {
		payload["hash"] = binaryHash
	}

	res, err := da.post("/init", payload)
	if err != nil {
		return nil, err
	}

	da.SessionToken = res.SessionToken
	return res, nil
}

// License authenticates with a license key.
func (da *DarkAuth) License(key string) (*Response, error) {
	if err := da.ensureSession(); err != nil {
		return nil, err
	}
	return da.post("/license", map[string]string{
		"session_token": da.SessionToken,
		"key":           strings.TrimSpace(key),
		"hwid":          da.HWID,
	})
}

// Login authenticates with username and password.
func (da *DarkAuth) Login(username, password string) (*Response, error) {
	if err := da.ensureSession(); err != nil {
		return nil, err
	}
	res, err := da.post("/login", map[string]string{
		"session_token": da.SessionToken,
		"username":      username,
		"password":      password,
		"hwid":          da.HWID,
	})
	if err != nil {
		return nil, err
	}
	return res, nil
}

// Register creates a new account with a license key.
func (da *DarkAuth) Register(username, password, key string) (*Response, error) {
	if err := da.ensureSession(); err != nil {
		return nil, err
	}
	res, err := da.post("/register", map[string]string{
		"session_token": da.SessionToken,
		"username":      username,
		"password":      password,
		"key":           strings.TrimSpace(key),
		"hwid":          da.HWID,
	})
	if err != nil {
		return nil, err
	}
	return res, nil
}

// Check verifies the current session is still valid.
func (da *DarkAuth) Check() bool {
	if da.SessionToken == "" {
		return false
	}
	res, err := da.post("/check", map[string]string{"session_token": da.SessionToken})
	if err != nil {
		return false
	}
	return res.Success
}

// GetVar retrieves a cloud variable value.
func (da *DarkAuth) GetVar(name string) (string, error) {
	if da.SessionToken == "" {
		return "", fmt.Errorf("not initialized")
	}
	res, err := da.post("/var/get", map[string]string{
		"session_token": da.SessionToken,
		"name":          name,
	})
	if err != nil {
		return "", err
	}
	return res.Value, nil
}

// SetVar sets a cloud variable value.
func (da *DarkAuth) SetVar(name, value string) (*Response, error) {
	if da.SessionToken == "" {
		return nil, fmt.Errorf("not initialized")
	}
	return da.post("/var/set", map[string]string{
		"session_token": da.SessionToken,
		"name":          name,
		"value":         value,
	})
}

// Log sends a log entry to the API.
func (da *DarkAuth) Log(message, level string) (*Response, error) {
	if da.SessionToken == "" {
		return nil, fmt.Errorf("not initialized")
	}
	if level == "" {
		level = "INFO"
	}
	return da.post("/log", map[string]string{
		"session_token": da.SessionToken,
		"message":       message,
		"level":         level,
	})
}

// ResetHWID requests a HWID reset for a key.
func (da *DarkAuth) ResetHWID(key string) (*Response, error) {
	if da.SessionToken == "" {
		return nil, fmt.Errorf("not initialized")
	}
	return da.post("/hwid/reset", map[string]string{
		"session_token": da.SessionToken,
		"key":           key,
	})
}

// GetChat retrieves chat messages from a channel.
func (da *DarkAuth) GetChat(channel string) ([]ChatMessage, error) {
	if da.SessionToken == "" {
		return nil, fmt.Errorf("not initialized")
	}
	if channel == "" {
		channel = "general"
	}
	res, err := da.get("/chat", map[string]string{
		"session_token": da.SessionToken,
		"channel":       channel,
	})
	if err != nil {
		return nil, err
	}
	return res.Messages, nil
}

// SendChat sends a chat message to a channel.
func (da *DarkAuth) SendChat(sender, message, channel string) (*Response, error) {
	if da.SessionToken == "" {
		return nil, fmt.Errorf("not initialized")
	}
	if channel == "" {
		channel = "general"
	}
	return da.post("/chat", map[string]string{
		"session_token": da.SessionToken,
		"channel":       channel,
		"sender":        sender,
		"message":       message,
	})
}

File: sdk/python/darkauth.py
"""
DARK-AUTH Official Python SDK
=============================
KeyAuth / Authly compatible client library.
Supports: Init, License Auth, User Login/Register, Cloud Variables, Chat, Logging, HWID lock.
"""

import hashlib
import platform
import subprocess
import uuid
import requests


class DarkAuth:
    def __init__(self, app_id: str, secret: str, api_url: str, version: str = "1.0.0"):
        self.app_id = app_id
        self.secret = secret
        self.version = version
        self.api_url = api_url.rstrip("/")
        self.session_token = None
        self.hwid = self.generate_hwid()
        self.user = None
        self.license_info = None

    @staticmethod
    def generate_hwid() -> str:
        """Collect stable hardware identifiers and return SHA-256 fingerprint."""
        components = [platform.node(), platform.machine(), platform.processor()]
        try:
            mac = hex(uuid.getnode())
            components.append(mac)
        except Exception:
            pass

        if platform.system() == "Windows":
            try:
                out = subprocess.check_output("wmic csproduct get uuid", shell=True).decode()
                lines = [l.strip() for l in out.splitlines() if l.strip()]
                if len(lines) > 1:
                    components.append(lines[1])
            except Exception:
                pass

        raw = ":".join(components)
        return hashlib.sha256(raw.encode()).hexdigest()

    @staticmethod
    def file_hash(file_path: str) -> str:
        """Calculate SHA-256 of file for binary integrity anti-tamper checking."""
        h = hashlib.sha256()
        with open(file_path, "rb") as f:
            while chunk := f.read(8192):
                h.update(chunk)
        return h.hexdigest()

    def _post(self, endpoint: str, data: dict) -> dict:
        url = f"{self.api_url}/{endpoint.lstrip('/')}"
        resp = requests.post(url, json=data, timeout=15)
        res_json = resp.json()
        if not res_json.get("success"):
            raise Exception(f"[{res_json.get('code', 'ERROR')}] {res_json.get('message', 'Request failed')}")
        return res_json

    def init(self, binary_path: str = None) -> dict:
        """Initialize session and check for software updates and anti-tamper hash."""
        h = self.file_hash(binary_path) if binary_path else None
        data = {
            "app_id": self.app_id,
            "secret": self.secret,
            "version": self.version,
            "hash": h,
        }
        res = self._post("/init", data)
        self.session_token = res.get("session_token")
        return res

    def license(self, key: str) -> dict:
        """Authenticate directly with a license key (KeyAuth style)."""
        if not self.session_token:
            self.init()
        data = {
            "session_token": self.session_token,
            "key": key.strip(),
            "hwid": self.hwid,
        }
        res = self._post("/license", data)
        self.license_info = res
        return res

    def login(self, username: str, password: str) -> dict:
        """Authenticate with existing user account."""
        if not self.session_token:
            self.init()
        data = {
            "session_token": self.session_token,
            "username": username,
            "password": password,
            "hwid": self.hwid,
        }
        res = self._post("/login", data)
        self.user = res.get("user")
        return res

    def register(self, username: str, password: str, key: str) -> dict:
        """Register a new account with a license key."""
        if not self.session_token:
            self.init()
        data = {
            "session_token": self.session_token,
            "username": username,
            "password": password,
            "key": key.strip(),
            "hwid": self.hwid,
        }
        res = self._post("/register", data)
        self.user = res.get("user")
        return res

    def check(self) -> bool:
        """Check if active session is still valid."""
        if not self.session_token:
            return False
        try:
            res = self._post("/check", {"session_token": self.session_token})
            return res.get("success", False)
        except Exception:
            return False

    def get_var(self, name: str) -> str:
        """Fetch remote cloud variable securely at runtime."""
        res = self._post("/var/get", {"session_token": self.session_token, "name": name})
        return res.get("value")

    def set_var(self, name: str, value: str) -> dict:
        """Modify user-writeable cloud variable."""
        return self._post("/var/set", {"session_token": self.session_token, "name": name, "value": value})

    def log(self, message: str, level: str = "INFO") -> dict:
        """Stream telemetry or error logs to the developer dashboard."""
        return self._post("/log", {"session_token": self.session_token, "message": message, "level": level})

    def reset_hwid(self, key: str) -> dict:
        """Request HWID reset for a key."""
        return self._post("/hwid/reset", {"session_token": self.session_token, "key": key})

    def get_chat(self, channel: str = "general") -> list:
        """Retrieve recent in-app chat or announcements."""
        url = f"{self.api_url}/chat?session_token={self.session_token}&channel={channel}"
        resp = requests.get(url, timeout=10)
        return resp.json().get("messages", [])

    def send_chat(self, sender: str, message: str, channel: str = "general") -> dict:
        """Broadcast an in-app chat message."""
        return self._post("/chat", {
            "session_token": self.session_token,
            "channel": channel,
            "sender": sender,
            "message": message,
        })
